src/Pulsar.Client/Internal/ProducerImpl.fs

```diff
@@ -142,6 +143,39 @@ type internal ProducerImpl<'T> private (producerConfig: ProducerConfiguration, c
             statTimer.Elapsed.Add(fun _ -> this.Mb.Post(Tick StatTick))
             statTimer.Start()
     
+    let cryptoTimer = new Timer()
+    let startCryptoTimer () =
+        if not (producerConfig.EncryptionKeys.Length = 0 || isNull producerConfig.CryptoKeyReader) then
+            cryptoTimer.Interval <- TimeSpan(hours = 4, minutes = 0, seconds = 0).TotalMilliseconds
+            cryptoTimer.AutoReset <- true
+            cryptoTimer.Elapsed.Add(fun _ -> this.Mb.Post(Tick UpdateEncryption))
+            cryptoTimer.Start()
+        
+    let initCrypto () =
+        if producerConfig.EncryptionKeys.Length = 0 || isNull(producerConfig.CryptoKeyReader) then
+            Log.Logger.LogDebug("{0} CryptoKeyReader not present, encryption not possible", prefix)
+            fun (_, payload) -> payload
+        else
+            let msgCrypto = MessageEncrypt(prefix, producerConfig.EncryptionKeys, producerConfig.CryptoKeyReader)
+            match producerConfig.ProducerCryptoFailureAction, msgCrypto.MayEncrypt with
+            | ProducerCryptoFailureAction.Send, false ->
+                Log.Logger.LogDebug("{0} Producer cannot encrypt message, messages will be sent in clear text", prefix)
+                fun (_, payload) -> payload
+            | ProducerCryptoFailureAction.Send, true ->
+                Log.Logger.LogDebug("{0} Producer will be encrypt all message", prefix)
+                // Nsec encrypt throw exceptions when payload size is greater than int.MaxValue
+                fun (msgMetadata: MessageMetadata, payload: byte []) ->
+                    msgCrypto.Encrypt(msgMetadata, payload)
+            | ProducerCryptoFailureAction.Fail, false ->
+                Log.Logger.LogDebug("{0} Producer cannot encrypt message", prefix)
+                fun (_, _) -> raise (CryptoException "Producer cannot encrypt message")
+            | ProducerCryptoFailureAction.Fail, true ->
+                Log.Logger.LogDebug("{0} Producer will be encrypt all message", prefix)
+                fun (msgMetadata: MessageMetadata, payload: byte []) ->
+                    msgCrypto.Encrypt(msgMetadata, payload)
```

Comment:
Wrong logs placement. Consider creating same function as java has.

Reply:
These logs depend on the settings of the producer and I decided that it is better to show them when creating the producer than each time a message is sent (the logs for each message will be the same)

Reply:
It's better to log each time also specifying the message id.
Also noticed, that 2 cases do exactly the same, you can group them to avoid code duplication